src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -0,0 +1,178 @@
+package io.iohk.ethereum.db.storage
+
+import akka.util.ByteString
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.db.storage.pruning.{PruningNodesKeyValueStorage, PruneResult, RangePrune}
+import io.iohk.ethereum.mpt.NodesKeyValueStorage
+import io.iohk.ethereum.rlp.RLPImplicitConversions.{toRlpList, _}
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{encode => rlpEncode, _}
+
+/**
+  * This class helps to deal with two problems regarding MptNodes storage:
+  * 1) Define a way to delete ones that are no longer needed
+  * 2) Avoids removal of nodes that can be used in diferent trie branches because the hash it's the same
+  *
+  * To deal with (1) when a node is no longer needed it's tagged with the corresponding block number in order to be
+  * able to release disk space used (as it's not going to be accessed any more)
+  *
+  * In order to solve (2), before saving a node, its wrapped with the number of references it has. The inverse operation
+  * is done when getting a node.
+  *
+  *
+  */
+class ReferenceCountNodeStorage(nodeStorage: NodeStorage, pruningOffset: BigInt, blockNumber: Option[BigInt] = None)
+  extends PruningNodesKeyValueStorage
+  with RangePrune {
+
+  import ReferenceCountNodeStorage._
+
+  override def get(key: ByteString): Option[NodeEncoded] = nodeStorage.get(key).map(_.toStoredNode).map(_.nodeEncoded.toArray)
+
+  override def update(toRemove: Seq[NodeHash], toUpsert: Seq[(NodeHash, NodeEncoded)]): NodesKeyValueStorage = {
+
+    require(blockNumber.isDefined)
+
+    val bn = blockNumber.get
+    val upsertChanges = toUpsert.foldLeft(Map.empty[NodeHash, StoredNode]) { (storedNodes, toUpsertItem) =>
+      val (nodeKey, nodeEncoded) = toUpsertItem
+      val storedNode: StoredNode = storedNodes.get(nodeKey)
+        .orElse(nodeStorage.get(nodeKey).map(_.toStoredNode))
+        .getOrElse(StoredNode(ByteString(nodeEncoded), 0))
+        .incrementReferences(1)
+
+      storedNodes + (nodeKey -> storedNode)
+    }
+
+    val toPruneInThisBlockKey = pruneKey(bn)
+    val pruneCandidates = nodeStorage.get(toPruneInThisBlockKey).map(_.toPruneCandidates).getOrElse(PruneCandidates(Nil))
```

Comment:
Minor detail: I'm not sure about storing the `pruneCandidates` in the `NodeStorage`, maybe we could create separate storage or rename `NodeStorage`?